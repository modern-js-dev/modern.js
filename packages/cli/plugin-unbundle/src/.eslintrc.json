{
  "rules": {
    "no-param-reassign": "off",
    "@typescript-eslint/no-loop-func": "off",
    "prefer-destructuring": "off",
    "consistent-return": "off",
    "max-statements": "off",
    "@typescript-eslint/no-shadow": "off",
    "max-depth": "off",
    "@typescript-eslint/member-ordering": "off",
    "max-lines": "off",
    "@typescript-eslint/no-base-to-string": "off",
    "@typescript-eslint/ban-ts-comment": "off",
    "callback-return": "off",
    "no-cond-assign": "off",
    "react-hooks/rules-of-hooks": "off"
  }
}
