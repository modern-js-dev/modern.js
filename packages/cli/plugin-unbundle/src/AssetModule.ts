import { path } from '@modern-js/utils';
import { cleanUrl } from './utils';

export const idToModules = new Map<string, AssetModule>();
export const fileToModules = new Map<string, AssetModule>();
export const urlToModules = new Map<string, AssetModule>();

export class AssetModule {
  // cleaned url
  url: string;

  // url with query or hash
  rawUrl: string;

  // resolved id from url
  _id: string;

  // absolute fs file path
  _filePath: string;

  // dependency's id
  dependencies: Set<string>;

  // depedent's id
  dependents: Set<string>;

  // import.hot.accept(deps, () => {}), deps id
  acceptIds: Set<string>;

  // import.hot.accept(() => {})
  selfAccepted: boolean;

  transformed: string | null;

  hmrTimestamp: number;

  etag: string;

  type: string;

  constructor(url: string) {
    this.rawUrl = url;
    this.url = cleanUrl(url);

    urlToModules.set(this.url, this);

    this._id = '';
    this._filePath = '';
    this.acceptIds = new Set<string>();
    this.dependencies = new Set<string>();
    this.dependents = new Set<string>();
    this.selfAccepted = false;
    this.transformed = null;
    this.hmrTimestamp = 0;
    this.etag = '';
    this.type = '';
  }

  get filePath() {
    return this._filePath;
  }

  set filePath(file: string) {
    this._filePath = file;
    this.type = path.extname(this._filePath);
    if (!fileToModules.has(file)) {
      fileToModules.set(file, this);
    }
  }

  get id() {
    return this._id;
  }

  set id(id) {
    this._id = id;
    if (!idToModules.has(id)) {
      idToModules.set(id, this);
    }
  }

  update() {
    // TODO
  }
}

export const createAssetModule = (url: string): AssetModule => {
  const existing = urlToModules.get(cleanUrl(url));

  if (existing) {
    return existing;
  }

  return new AssetModule(url);
};

// invalidate transform cache
// mark as been hmr update by timestamp
export const invalidateAssetModule = (
  assetModule: AssetModule,
  timestamp: number,
) => {
  assetModule.hmrTimestamp = timestamp;
  assetModule.transformed = null;
};
