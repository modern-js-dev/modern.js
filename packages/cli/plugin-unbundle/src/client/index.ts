// css HMR
const styles = new Map();

export const removeStyle = id => {
  if (styles.has(id)) {
    document.head.removeChild(styles.get(id));
    styles.delete(id);
  }
};

export const updateStyle = (id, code) => {
  const style = styles.get(id);
  if (!style) {
    const styleEl = document.createElement('style');
    const codeEl = document.createTextNode(code);
    styleEl.type = 'text/css';
    styleEl.appendChild(codeEl);
    document.head.appendChild(styleEl);
    styles.set(id, styleEl);
  } else {
    style.innerHTML = code;
  }
};

const ERROR_OVERLAY_NAME = 'modern-js-error-overlay';

// check if has  error overlay
function hasErrorOverlay() {
  return document.querySelectorAll(ERROR_OVERLAY_NAME).length;
}
// clear react error overlay
function clearErrorOverlay() {
  document.querySelectorAll(ERROR_OVERLAY_NAME).forEach(el => el.remove());
}

// create react error overlay
function createErrorOverlay(data) {
  const ErrorOverlay = customElements.get(ERROR_OVERLAY_NAME);
  if (ErrorOverlay) {
    const overlay = new ErrorOverlay(data);
    clearErrorOverlay();
    document.body.appendChild(overlay);
  }
}

type LoggerFunction = (...args: any[]) => void;
type LoggerType = {
  info: LoggerFunction;
  error: LoggerFunction;
  warn: LoggerFunction;
};
const logger = [`info`, 'error', 'warn'].reduce((m, c) => {
  m[c] = (...params) =>
    // eslint-disable-next-line no-console
    console[c === 'info' ? 'log' : c](`[HMR]:`, ...params);
  return m;
}, {} as LoggerType);

let isFisrtUpdate = true;

const handleMessage = event => {
  const data = JSON.parse(event.data);

  switch (data.type) {
    case 'connected': {
      logger.info(`Connected`);
      // TODO: socket ping
      break;
    }
    case 'reload': {
      window.location.reload();
      break;
    }
    case 'update': {
      // if there is already a error in first update
      // which meanings some resource load failed
      // so should reload page to ensure app run correctly
      if (isFisrtUpdate && hasErrorOverlay()) {
        window.location.reload();
        return;
      }

      isFisrtUpdate = false;

      try {
        const timestamp = data.hmrTimestamp;
        for (const { id, accepted } of data.changes) {
          doAccept(id, accepted, timestamp);
        }
        clearErrorOverlay();
      } catch (err) {
        logger.error('Hot Update Error', err);
        createErrorOverlay({
          title: 'Hot Update Error',
          loc: {
            file: data.url || '',
          },
          message: err.message,
          stack: err.stack,
        });
      }

      break;
    }
    // handle removed dependencies after hmr update
    // eg: remove style tags in head element
    case 'prune': {
      for (const id of data.changes) {
        doPrune(id);
      }
      break;
    }
    case 'error': {
      logger.error(data.message);
      createErrorOverlay(data);
      break;
    }
    case 'close': {
      logger.info(`Disconnected!`);
      break;
    }
    default: {
      logger.error(`Invalid message type:`, data);
    }
  }
};

const handleError = err => {
  logger.error(`connection error`, err);
};

let retries = 0;
const handleClose = () => {
  logger.info(`Disconnected!`);
  if (retries <= 10) {
    retries += 1;
    setTimeout(() => {
      initSocket();
    }, 1000 * Math.pow(2, retries) + Math.random() * 100);
  }
};

// init websocket client connection
const initSocket = () => {
  const ws = new WebSocket(
    `${location.origin.replace(/http/, 'ws')}/_modern_js_hmr_ws`,
  );
  ws.onerror = handleError;
  ws.onmessage = handleMessage;
  ws.onclose = handleClose;
};

initSocket();

const strip = url => url.replace(/(\?.*$|#.*$)/, '');

const noop = function () {
  /** */
};

const modules = new Map();

let updateQueue = [];
let updating = false;

const doAccept = async (id, acceptedPath, timestamp) => {
  if (!updateQueue.includes(id)) {
    updateQueue.push({ id, acceptedPath });
  }

  if (!updating) {
    updating = true;
    await Promise.resolve();
    updating = false;
    const loading = updateQueue.map(({ id: path, acceptedPath: accepted }) =>
      update(path, accepted, timestamp),
    );
    updateQueue = [];
    await Promise.all(loading);
  }
};

const update = async (id, acceptedPath, timestamp) => {
  if (!modules.has(id)) {
    return;
  }

  const mod = modules.get(id);
  let callbacks = [];
  const modulesToUpdate = new Set();
  if (mod._selfAccepted) {
    modulesToUpdate.add(id);
    callbacks = [mod._selfAccepted];
  } else {
    modulesToUpdate.add(id);
    mod._acceptedCallbacks.forEach(({ deps, callback }) => {
      if (deps.some(dep => dep === acceptedPath)) {
        modulesToUpdate.add(acceptedPath);
        callbacks.push(callback);
      }
    });
  }

  const loadedModules = await Promise.all(
    Array.from(modulesToUpdate).map(m => {
      const newUrl = __addQuery__(m, `t=${timestamp}`);
      return import(newUrl);
    }),
  );

  for (const callback of callbacks) {
    typeof callback === 'function' && callback(loadedModules);
  }

  logger.info(`hot updated: accepted via ${id}`);
};

/** Called when a new module is loaded, to run cleanup on the old module (if needed) */
const doDispose = id => {
  const mod = modules.get(id);
  const callbacks = mod._disposeCallbacks;

  callbacks.map(callback => callback());
};

const doPrune = id => {
  if (modules.has(id)) {
    const mod = modules.get(id);
    const callbacks = mod._prunCallbacks;
    callbacks.map(callback => callback());
    modules.delete(id);
  }
};

export const createHotContext = url => {
  const id = strip(url);
  const mod = modules.get(id);

  if (mod) {
    // mod.lock();
    doDispose(mod._id);
    mod._disposeCallbacks = [];
    mod._acceptedCallbacks = [];
    mod._selfAccepted = false;
    return mod;
  }

  const hot = {
    // private
    _id: id,
    _declinedDeps: {},
    _acceptedDeps: {},
    _selfAccepted: false,
    _selfDeclined: false,
    _disposeCallbacks: [],
    _acceptedCallbacks: [],
    _prunCallbacks: [],
    // hot API
    accept(dep, callback) {
      if (dep === undefined) {
        hot._selfAccepted = true;
      } else if (typeof dep === 'function') {
        hot._selfAccepted = dep;
      } else if (Array.isArray(dep)) {
        hot._acceptedCallbacks.push({ deps: dep, callback: callback || noop });
      } else {
        hot._acceptedCallbacks.push({
          deps: [dep],
          callback: callback || noop,
        });
      }
    },
    decline(dep) {
      if (dep === undefined) {
        hot._selfDeclined = true;
      } else if (Array.isArray(dep)) {
        for (const d of dep) {
          hot._acceptedDeps[d] = true;
        }
      } else {
        hot._declinedDeps[dep] = true;
      }
    },
    dispose(callback) {
      hot._disposeCallbacks.push(callback);
    },
    prune(callback) {
      hot._prunCallbacks.push(callback);
    },
    invalidate() {
      window.location.reload();
    },
  };

  modules.set(id, hot);

  return hot;
};

export const __addQuery__ = (url, query) =>
  /\?.*/.test(url) ? `${url}&${query}` : `${url}?${query}`;
