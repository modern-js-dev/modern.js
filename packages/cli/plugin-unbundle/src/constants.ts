import path from 'path';

export const MODE = 'development';

export const WEB_MODULES_DIR = './node_modules/.modern_js_web_modules';

export const META_DATA_FILE_NAME = 'metadata.json';

export const DEFAULT_DEPS: string[] = [
  'react',
  'react-dom',
  '@modern-js/runtime',
  '@modern-js/runtime/loadable',
  '@modern-js/runtime/model',
  '@modern-js/runtime/styled',
  '@modern-js/runtime/head',
  '@modern-js/runtime/model',
  '@modern-js/runtime/plugins',
  '@modern-js/runtime/router',
  '@modern-js/runtime/ssr',
  '@modern-js/create-request',
];

export const MODERN_JS_INTERNAL_PACKAGES: Record<string, string> = {
  '@modern-js/plugin-state': '@modern-js/runtime',
  '@modern-js/plugin-router': '@modern-js/runtime',
  '@modern-js/plugin-ssr': '@modern-js/runtime',
  '@modern-js/create-request/client': '@modern-js/plugin-unbundle',
};

export const VIRTUAL_DEPS_MAP: Record<string, string> = {
  '@modern-js/create-request': `export * from '@modern-js/create-request/client';`,
  '@modern-js/runtime/plugins': `
    export { default as router } from '@modern-js/plugin-router';
    export { default as ssr } from '@modern-js/plugin-ssr';
    export { default as state } from '@modern-js/plugin-state';
  `,
  '@modern-js/runtime/router': `export * from '@modern-js/plugin-router';`,
  '@modern-js/runtime/model': `export * from '@modern-js/plugin-state';`,
  '@modern-js/runtime/ssr': `export * from '@modern-js/plugin-ssr';`,
  domino: 'export {}; export default {};',
};

export const HOST = '0.0.0.0';

// [@\w] - Match a word-character or @ (valid package name)
// (?!.*(:\/\/)) - Ignore if previous match was a protocol (ex: http://)
export const BARE_SPECIFIER_REGEX = /^[@\w](?!.*(:\/\/))/;

export const DEFAULT_EXTENSIONS = [
  'web.mjs',
  'mjs',
  'web.js',
  'js',
  'web.ts',
  'ts',
  'web.tsx',
  'tsx',
  'json',
  'web.jsx',
  'jsx',
].map(ext => `.${ext}`);

export const CSS_REGEX = /\.(css|sass|scss|less)($|\?)/;
export const CSS_MODULE_REGEX = /\.module.(css|sass|scss|less)($|\?)/;

// hmr sockjs client, should inject modern build into client
export const DEV_CLIENT_PATH = path.resolve(__dirname, '../modern/client');

export const DEV_CLIENT_PATH_ALIAS = `@_modern_js_dev_client`;

export const DEV_CLIENT_URL = `/${DEV_CLIENT_PATH_ALIAS}/index.js`;

export const ASSETS_REGEX =
  /\.(gif|png|jpe?g|webp|bmp|ico|svg|woff|woff2|eot|ttf|otf|ttc)$/;

export const IS_DISABLE_REACT_REFRESH = process.env.FAST_REFRESH === 'false';

export const CSS_URL_FUNCTION_REGEX =
  /url\(\s*('[^']+?'|"[^"]+?"|[^'"]+?)\s*\)/g;

export const GLOBAL_CACHE_DIR_NAME = '__modern_js_web_modules__';

// cjs node_modules to esm node_modules
export const TEMP_MODULES_DIR = './node_modules/.cache/modern_js_modules';

// babel macros extensions
export const BABEL_MACRO_EXTENSIONS = '.macro';

// api function file query
export const LAMBDA_API_FUNCTION_QUERY = '__lambda_api_function__';

// dependencies that should be ignored in deps optimizaion
// can't convert to esm format
export const IGNORE_OPTIMIZE_DEPS = ['domino'];

// default lazy import ui components library
export const DEFAULT_LAZY_IMPORT_UI_COMPONENTS = [
  'antd',
  '@arco-design/web-react',
];

export const ESBUILD_RESOLVE_PLUGIN_NAME = 'esm-resolve-plugin';

export const BFF_API_DIR = './api';

export const DEFAULT_PDN_HOST = `pdn.zijieapi.com`;
