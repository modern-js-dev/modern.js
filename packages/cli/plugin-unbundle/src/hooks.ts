import { createAsyncWaterfall } from '@modern-js/plugin';

export interface UnbundleDependencies {
  defaultDeps: string[];
  internalPackages: Record<string, string>;
  virtualDeps: Record<string, string>;
  defaultPdnHost: string;
}

export const unbundleDependencies =
  createAsyncWaterfall<UnbundleDependencies>();

export const hooks = {
  unbundleDependencies,
};

declare module '@modern-js/core' {
  export interface Hooks {
    unbundleDependencies: typeof unbundleDependencies;
  }
}
