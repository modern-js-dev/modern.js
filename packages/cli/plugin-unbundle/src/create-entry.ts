import path from 'path';
import {
  fs,
  isFastRefresh,
  createDebugger,
  isTypescript,
  readTsConfig,
  generateMetaTags,
  getEntryOptions,
} from '@modern-js/utils';
import { IAppContext, NormalizedConfig } from '@modern-js/core';
import type { Entrypoint } from '@modern-js/types';
import { DEV_CLIENT_PATH_ALIAS, DEV_CLIENT_URL } from './constants';

const debug = createDebugger('esm:create-entry');

export interface BaseTemplateVariables {
  meta: string;
  title: string;
  staticPrefix: string;
  mountId: string;
  [param: string]: string;
}

export interface TemplateVariables extends BaseTemplateVariables {
  topTemplate: string;
  headTemplate: string;
  bodyTemplate: string;
}

export interface TemplatesMap {
  index: boolean | string;
  top: boolean | string;
  head: boolean | string;
  body: boolean | string;
  bottom: boolean | string;
}

// inject entry js script and react fast refresh runtime code and error-overlay client
const injectScripts = (
  entrypoint: Entrypoint,
  appDirectory: string,
): string => {
  const pathname = path.relative(appDirectory, entrypoint.entry);

  debug(`inject-scripts src: ${pathname}`);

  const scripts = [
    `<script type="module" src="/${pathname}"></script>`,
    `<script type="module" src="/${DEV_CLIENT_PATH_ALIAS}/error-overlay.js"></script>\n<script type="module" src="${DEV_CLIENT_URL}"></script>`,
  ];

  if (isFastRefresh()) {
    const runtimePath = require.resolve(
      'react-refresh/cjs/react-refresh-runtime.development.js',
    );

    const reactRefreshCode = fs
      .readFileSync(runtimePath, { encoding: 'utf-8' })
      .replace(`process.env.NODE_ENV`, JSON.stringify('development'));

    scripts.push(`<script>
    function debounce(e,t){let u;return()=>{clearTimeout(u),u=setTimeout(e,t)}}
    {
      const exports = {};
      ${reactRefreshCode}
      exports.performReactRefresh = debounce(exports.performReactRefresh, 30);
      window.$RefreshRuntime$ = exports;
      window.$RefreshRuntime$.injectIntoGlobalHook(window);
      window.$RefreshReg$ = () => {};
      window.$RefreshSig$ = () => (type) => type;
    }
  </script>`);
  }

  return scripts.join('\n');
};

// provide process/module global variable
// TODO: should take better strategy
const injectEnv = (userConfig: NormalizedConfig): string => {
  // inject globalVars
  let {
    source: { globalVars },
  } = userConfig;

  globalVars = globalVars || {};

  const globalVarKeys = Object.keys(globalVars);

  let gloablVarStr = ``;

  globalVarKeys.forEach(key => {
    gloablVarStr += `window.${key}=${JSON.stringify(globalVars![key])}\n`;
  });

  return `<script>${gloablVarStr}\nwindow.process={env: { NODE_ENV: 'development'}};\nwindow.module={};\n</script>\n`;
};

// inject const enum value to global object
const injectConstEnums = (appDirectory: string): string => {
  if (isTypescript(appDirectory)) {
    // read tsconfig.json
    const tsconfigJSON = readTsConfig(appDirectory);

    // .d.ts files which need to be
    const ts = require('typescript');
    const {
      raw: { include },
    } = ts.parseJsonConfigFileContent(tsconfigJSON, ts.sys, appDirectory);

    if (include?.length) {
      const { parseDTS } = require('./parse-dts');
      const enums = parseDTS(
        include
          .filter((file: string) => path.extname(file))
          .map((file: string) => path.resolve(appDirectory, file)),
      );

      return `<script>${Object.keys(enums)
        .map((key: string) => `window.${key}=${JSON.stringify(enums[key])};`)
        .join('\n')}</script>`;
    }
  }

  return '';
};

const renderTemplate = (
  filepath: string,
  variables: TemplateVariables | BaseTemplateVariables,
) => {
  const content = fs.readFileSync(filepath, 'utf8');

  return require('lodash.template')(content)(variables);
};

const initTemplateVariables = (
  userConfig: NormalizedConfig,
  entryName: string,
): BaseTemplateVariables => {
  const {
    output: {
      title,
      titleByEntries,
      meta,
      metaByEntries,
      templateParameters,
      templateParametersByEntries,
      mountId,
      assetPrefix,
    },
  } = userConfig;

  const titleVariable = getEntryOptions(entryName, title, titleByEntries);

  const metaVariable = generateMetaTags(
    getEntryOptions(entryName, meta, metaByEntries),
  );

  return {
    staticPrefix: assetPrefix || '/',
    title: titleVariable,
    meta: metaVariable,
    mountId: mountId!,
    // TODO: favicon
    ...getEntryOptions<Record<string, unknown> | undefined>(
      entryName,
      templateParameters,
      templateParametersByEntries,
    ),
  };
};

const createEntryHtml = (
  appContext: IAppContext,
  userConfig: NormalizedConfig,
  entryName: string,
) => {
  const templateVariables = initTemplateVariables(userConfig, entryName);

  const template = appContext.htmlTemplates[entryName];

  const output = renderTemplate(template, { ...templateVariables });

  fs.outputFileSync(template, output);
};

export const createHtml = (
  userConfig: NormalizedConfig,
  appContext: IAppContext,
) => {
  appContext.entrypoints.forEach(({ entryName }) => {
    createEntryHtml(appContext, userConfig, entryName);
  });
};

export const createHtmlPartials = (
  entrypoint: Entrypoint,
  appContext: IAppContext,
  config: NormalizedConfig,
) =>
  [
    injectScripts(entrypoint, appContext.appDirectory),
    injectEnv(config),
    injectConstEnums(appContext.appDirectory),
  ].join('\n');
