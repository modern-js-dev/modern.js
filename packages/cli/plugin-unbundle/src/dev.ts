import { chalk, logger } from '@modern-js/utils';
import { IAppContext, NormalizedConfig } from '@modern-js/core';

export const startTimer = {
  start: 0,
  end: 0,
};

export const dev = async (
  config: NormalizedConfig,
  appContext: IAppContext,
) => {
  logger.log(chalk.cyanBright(`Starting the development server...\n`));

  startTimer.start = Date.now();

  const [{ startDevServer }, { createHtml }] = await Promise.all([
    import('./server'),
    import('./create-entry'),
  ]);

  // create html template for each js entry
  createHtml(config, appContext);

  const closeServer = await startDevServer(config, appContext);
  return closeServer;
};
