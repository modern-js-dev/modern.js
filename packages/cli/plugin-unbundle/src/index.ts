import type { CliPlugin } from '@modern-js/core';
import { PLUGIN_SCHEMAS } from '@modern-js/utils';
import { dev } from './dev';
import { hooks } from './hooks';

export * from './hooks';

export default (): CliPlugin => ({
  name: '@modern-js/plugin-unbundle',

  registerHook: hooks,

  setup: api => {
    let closeDevServer: (() => Promise<void>) | undefined;

    return {
      validateSchema() {
        return PLUGIN_SCHEMAS['@modern-js/plugin-unbundle'];
      },
      commands({ program }) {
        const appContext = api.useAppContext();
        const config = api.useResolvedConfigContext();

        const devCommand = program.commandsMap.get('dev');

        devCommand?.option('--unbundled', 'dev with unbundled mode');

        if (
          process.argv.slice(2)[0] === 'dev' &&
          process.argv.includes('--unbundled')
        ) {
          devCommand?.action(async () => {
            if (closeDevServer) {
              await closeDevServer();
              closeDevServer = undefined;
            }
            closeDevServer = await dev(api, config, appContext);
          });
        }
      },
      async beforeRestart() {
        if (closeDevServer) {
          await closeDevServer();
        }
        closeDevServer = undefined;
      },
      async htmlPartials({ entrypoint, partials }) {
        if (process.argv[2] === 'dev' && process.argv.includes('--unbundled')) {
          const appContext = api.useAppContext();
          const config = api.useResolvedConfigContext();

          const { createHtmlPartials } = await import('./create-entry');

          partials.head.push(
            createHtmlPartials(entrypoint, appContext, config),
          );
        }

        return {
          entrypoint,
          partials,
        };
      },
    };
  },
});
