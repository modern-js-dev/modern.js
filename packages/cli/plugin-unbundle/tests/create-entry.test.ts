import { injectEnv } from '../src/create-entry';

describe('inject env', () => {
  it('should inject globalVars correctly', () => {
    expect(
      injectEnv({
        source: {
          globalVars: {
            'process.env.foo': 'foo',
          },
        },
      } as any),
    ).toMatchSnapshot();
  });
});
