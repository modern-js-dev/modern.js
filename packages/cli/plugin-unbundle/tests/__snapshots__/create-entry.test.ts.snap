// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inject env should inject globalVars correctly 1`] = `
"<script>
window.process={env: { NODE_ENV: 'development'}};;
window.module={};
window.process.env.foo=\\"foo\\"

</script>
"
`;
